### DISCUSSION TITLE
Tabulating numerical gradients

### DISCUSSION THREAD

**Discussion by guro86** [snl-dakota/dakota#42] [Q&A] [answered]
Hello all,

I have a simple question. I am using the sampling method to generate LHS input samples and corresponding predictions from a legacy Fortran code. I am also interested in developing and tabulating the numerical gradients associated with the predictions. So I turned on numerical gradients, and Dakota calculated and presented those in the standard output.

Could one also tabulate the values in any straightforward way? In the same way, the predictions are tabulated in the dakota_tabular.dat file.

I appreciate any input on this. 

Thanks,

**Comment by briadam** [comment, marked as answer]
Dakota does not offer a direct way to export derivative information to a tabular file. A perhaps better way than scraping the console output for the derivative information would be to extract the data from Dakota's HDF5 output: https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/output/hdf/organizationofevaluations.html

If you didn't generate HDF5 output when first running Dakota, you can add the HDF5 output options https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/reference/environment-results_output.html, and then restart: https://snl-dakota.github.io/docs/6.18.0/users/usingdakota/running/restart.html

Brian

**Reply by guro86** [reply]
I got it, thanks. I will then try to compile Dakota with hdf5 support. Thanks.

**Reply by guro86** [reply]
Hello again, I did a toy implementation, and this worked fine. Now that I am using the real code-coupling, it reports nan values for the gradients in the HDF-file. However, in the output, gradients are written. I appreciate any ideas you may have to resolve this. The input file is attached below:

environment
  
  tabular_data 
  results_output hdf
  results_output_file 'dakota'
  output_file 'dakota.out'
  error_file 'dakota.err'

method 

  sampling samples 4

variables 

  uniform_uncertain 5
  descriptors 'diff' 'ath' 'gb_saturation' 'gb_sweeping' 'crack'
  lower_bounds 0.1 .1 .1 .1 0 
  upper_bounds 40  10 10 1 1

model

  single 
  id 'fgr'


responses

  response_functions 31
  numerical_gradients 
  no_hessians 

  descriptors 
  'HS2'
  'TR1'
  'DR1'
  'HR4'
  'HS1'
  'LS3'
  'LS2'
  'HR5'
  'BR1'
  'LR1'
  'HR2'
  'PW3_3'
  'PK1_3'
  'PK2_2'
  'PK2_4'
  'PK6_4'
  'PW3_2'
  'PK2_3'
  'PK1_1'
  'PK2_1'
  'PK1_2'
  'PK6_3'
  'PK1_4'
  'G20_2'
  'G20_4'
  'G20_1'
  'A20_1'
  'A20_3'
  'A20_4'
  'A10_3'
  'A10_4'

interface 

  fork
  analysis_driver 'driver.py'

  link_files 'files.yaml' 'calibration_template.inp'

  work_dir named 'workdir/run'

  parameters_file 'params.in'
  results_file 'results.out'

  labeled

  directory_tag

**Reply by jadamstephens** [reply]
My guess is, you're looking at the evaluation data in /interfaces. Try looking in /models. Because you told Dakota to estimate gradients numerically, the interface evaluation data will contain only function values. A Dakota model uses these function values to estimate the gradients, and so it's the model that records the gradients as part of its evaluation data.

Adam
